File: examples/statsd.erl
%%
%% Example erlang client implementation for the Etsy statsd
%%
%% Copyright (c) 2011-2012 Daniel Schauenberg, Sean Johnson
%%
%% Permission is hereby granted, free of charge, to any person
%% obtaining a copy of this software and associated documentation
%% files (the "Software"), to deal in the Software without
%% restriction, including without limitation the rights to use,
%% copy, modify, merge, publish, distribute, sublicense, and/or sell
%% copies of the Software, and to permit persons to whom the
%% Software is furnished to do so, subject to the following
%% conditions:
%%
%% The above copyright notice and this permission notice shall be
%% included in all copies or substantial portions of the Software.
%%
%% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
%% EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
%% OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
%% NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
%% HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
%% WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
%% FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
%% OTHER DEALINGS IN THE SOFTWARE.
%%
%% Implements the following functions:
%%   increment(Stat) -> Increment the counter for the Stat by 1 with a 1.0 Sample Rate.
%%   increment(Stat, Amount) -> Increment the counter by the Amount for the Stat with a 1.0 Sample Rate.
%%   increment(Stat, Amount, SampleRate) -> Increment the counter for Stat by the Amount with the Sample Rate.
%%   decrement(Stat) -> Decrement the counter for the Stat by 1 with a 1.0 Sample Rate.
%%   decrement(Stat, Amount) -> Decrement the counter by the Amount for the Stat with a 1.0 Sample Rate.
%%   decrement(Stat, Amount, SampleRate) -> Decrement the counter for Stat by the Amount with the Sample Rate.
%%   timing(Stat, Time) -> Record a Time for the Stat with a 1.0 Sample Rate.
%%   timing(Stat, Time, SampleRate) -> Record Time for the Stat with the Sample Rate.
%%   gauge(Stat, Reading) -> Record a Reading of the Stat
%%
%% Run tests with:
%%   Run statsd with the parameters set in the -defines()
%%   % erlc -pa /path/to/eunit/ebin -DTEST -o.statsd.erl
%%   % erl -noshell -pa. -eval "eunit:test(statsd, [verbose])" -s init stop
%%
%%
-module(statsd).
% Meta information
-author("Daniel Schauenberg <d@unwiredcouch.com>").
-author("Sean Johnson <sean@talkto.com>").
% defines
-define(HOST, "localhost").
-define(PORT, 8125).
% exports
-export([increment/1,increment/2,increment/3,
         decrement/1, decrement/2, decrement/3,
         timing/2, timing/3,
         gauge/2]).

%
% API functions
%

% functions for incrementing counters
increment(Stat) -> increment(Stat, 1, 1.0).
increment(Stat, Delta) when is_integer(Delta) ->
  increment(Stat, Delta, 1.0);
increment(Stat, Samplerate) when is_float(Samplerate) ->
  increment(Stat, 1, Samplerate).
increment(Stat, Delta, Samplerate) when is_integer(Delta), is_float(Samplerate) ->
  send({counter, Stat, Delta, Samplerate}).

% functions for decrementing counters
decrement(Stat) -> decrement(Stat, 1, 1.0).
decrement(Stat, Delta) when is_integer(Delta) ->
  decrement(Stat, Delta, 1.0);
decrement(Stat, Samplerate) when is_float(Samplerate) ->
  decrement(Stat, 1, Samplerate).
decrement(Stat, Delta, Samplerate) when is_integer(Delta), is_float(Samplerate) ->
  send({counter, Stat, -abs(Delta), Samplerate}).

% functions for timing values
timing(Stat, Time) when is_float(Time); is_integer(Time) ->
  timing(Stat, Time, 1.0).
timing(Stat, Time, Samplerate) when is_float (Time); is_integer(Time),
                                    is_float(Samplerate) ->
  send({timer, Stat, Time, Samplerate}).

% functions for guages
gauge(Stat, Reading) ->
  send({gauge, Stat, Reading}).

%
% update functions
%

% recursive functions for multiple stats
send({counter, [H|Stats], Delta, Samplerate}) ->
  send({counter, H, Delta, Samplerate}),
  send({counter, Stats, Delta, Samplerate});
send({counter, [], _, _}) -> {ok, "Success."};

send({timer, [H|Stats], Delta, Samplerate}) ->
  send({timer, H, Delta, Samplerate}),
  send({timer, Stats, Delta, Samplerate});
send({timer, [], _, _}) -> {ok, "Success."};

send({gauge, [H|Stats], Reading}) ->
  send({gauge, H, Reading}),
  send({gauge, Stats, Reading});
send({gauge, [], _}) -> {ok, "Success."};

% functions for single stats
send({counter, Stat, Delta, Samplerate}) when Samplerate < 1.0, is_atom(Stat) ->
  Rand = random:uniform(),
  if
    Rand =< Samplerate ->
      send_udp_message(?HOST, ?PORT, io_lib:format("~p:~p|c|@~p",
                                                   [Stat, Delta, Samplerate]));
    true -> {ok, "Success."}
  end;

send({counter, Stat, Delta, _}) ->
  send_udp_message(?HOST, ?PORT, io_lib:format("~p:~p|c", [Stat, Delta]));

send({timer, Stat, Time, Samplerate}) when Samplerate < 1.0 ->
  Rand = random:uniform(),
  if
    Rand =< Samplerate ->
      send_udp_message(?HOST, ?PORT, io_lib:format("~p:~p|ms|@~p",
                                                   [Stat, Time, Samplerate]));
    true -> {ok, "Success."}
  end;

send({timer, Stat, Time, _}) ->
  send_udp_message(?HOST, ?PORT, io_lib:format("~p:~p|ms", [Stat, Time]));

send({gauge, Stat, Reading}) ->
  send_udp_message(?HOST, ?PORT, io_lib:format("~p:~p|g", [Stat, Reading])).

%
% Raw UDP send message function
%
send_udp_message(Host, Port, Msg) when is_integer(Port),
                                         is_list(Host) ->
  {ok, Socket} = gen_udp:open(0, [binary]),
  ok = gen_udp:send(Socket, Host, Port, Msg),
  gen_udp:close(Socket),
  {ok, "Success."}.

%
% unit tests
%
% EUnit headers
-ifdef(TEST).
-include_lib("eunit/include/eunit.hrl").
% test functions go here
increment_test_() ->
  [?_assert(increment(stat1) =:= {ok, "Success."}),
   ?_assert(increment(stat1, 1) =:= {ok, "Success."}),
   ?_assert(increment(stat1, 1, 0.5) =:= {ok, "Success."}),
   ?_assert(increment([stat1, stat2]) =:= {ok, "Success."}),
   ?_assert(increment([stat1, stat2], 1) =:= {ok, "Success."}),
   ?_assert(increment([stat1, stat2], 1, 0.5) =:= {ok, "Success."})
       ].
decrement_test_() ->
  [?_assert(decrement(stat1) =:= {ok, "Success."}),
   ?_assert(decrement(stat1, 1) =:= {ok, "Success."}),
   ?_assert(decrement(stat1, -1) =:= {ok, "Success."}),
   ?_assert(decrement(stat1, 1, 0.5) =:= {ok, "Success."}),
   ?_assert(decrement([stat1, stat2]) =:= {ok, "Success."}),
   ?_assert(decrement([stat1, stat2], 1) =:= {ok, "Success."}),
   ?_assert(decrement([stat1, stat2], -1) =:= {ok, "Success."}),
   ?_assert(decrement([stat1, stat2], 1, 0.5) =:= {ok, "Success."})
       ].
timing_test_() ->
  [
   ?_assert(timing(stat1, 100) =:= {ok, "Success."}),
   ?_assert(timing(stat1, 100, 0.5) =:= {ok, "Success."}),
   ?_assert(timing([stat1, stat2], 100) =:= {ok, "Success."}),
   ?_assert(timing([stat1, stat2], 100, 0.5) =:= {ok, "Success."})
       ].
gauge_test_() ->
 [
  ?_assert(gauge(stat1, 100) =:= {ok, "Success."}),
  ?_assert(gauge([stat1, stat2], 100) =:= {ok, "Success."})
      ].

-endif.

File: examples/StatsD.scala
/*

Scala implementation of Andrew Gwozdziewycz's StatsdClient.java

Copyright (c) 2013 Joshua Garnett

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

package com.statsd

import java.io.IOException
import java.net._
import java.nio.ByteBuffer
import java.nio.channels.DatagramChannel
import java.util.Random
import org.slf4j.LoggerFactory
import akka.actor._

/**
 * Client for sending stats to StatsD uses Akka to manage concurrency
 *
 * @param context The Akka ActorContext
 * @param host The statsd host
 * @param port The statsd port
 * @param multiMetrics If true, multiple stats will be sent in a single UDP packet
 * @param packetBufferSize If multiMetrics is true, this is the max buffer size before sending the UDP packet
 */
class StatsD(context: ActorContext,
             host: String,
             port: Int,
             multiMetrics: Boolean = true,
             packetBufferSize: Int = 1024) {

  private val rand = new Random()

  private val actorRef = context.actorOf(Props(new StatsDActor(host, port, multiMetrics, packetBufferSize)))

  /**
   * Sends timing stats in milliseconds to StatsD
   *
   * @param key name of the stat
   * @param value time in milliseconds
   */
  def timing(key: String, value: Int, sampleRate: Double = 1.0) = {
    send(key, value.toString, StatsDProtocol.TIMING_METRIC, sampleRate)
  }

  /**
   * Decrement StatsD counter
   *
   * @param key name of the stat
   * @param magnitude how much to decrement
   */
  def decrement(key: String, magnitude: Int = -1, sampleRate: Double = 1.0) = {
    increment(key, magnitude, sampleRate)
  }

  /**
   * Increment StatsD counter
   *
   * @param key name of the stat
   * @param magnitude how much to increment
   */
  def increment(key: String, magnitude: Int = 1, sampleRate: Double = 1.0) = {
    send(key, magnitude.toString, StatsDProtocol.COUNTER_METRIC, sampleRate)
  }

  /**
   * StatsD now also supports gauges, arbitrary values, which can be recorded.
   *
   * @param key name of the stat
   * @param value Can be a fixed value or increase or decrease (Ex: "10" "-1" "+5")
   */
  def gauge(key: String, value: String = "1", sampleRate: Double = 1.0) = {
    send(key, value, StatsDProtocol.GAUGE_METRIC, sampleRate)
  }

  /**
   * StatsD supports counting unique occurrences of events between flushes, using a Set to store all occurring events.
   *
   * @param key name of the stat
   * @param value value of the set
   */
  def set(key: String, value: Int, sampleRate: Double = 1.0) = {
    send(key, value.toString, StatsDProtocol.SET_METRIC, sampleRate)
  }

  /**
   * Checks the sample rate and sends the stat to the actor if it passes
   */
  private def send(key: String, value: String, metric: String, sampleRate: Double): Boolean = {
    if (sampleRate >= 1 || rand.nextDouble <= sampleRate) {
      actorRef ! SendStat(StatsDProtocol.stat(key, value, metric, sampleRate))
      true
    }
    else {
      false
    }
  }
}

object StatsDProtocol {
  val TIMING_METRIC = "ms"
  val COUNTER_METRIC = "c"
  val GAUGE_METRIC = "g"
  val SET_METRIC = "s"

  /**
   * @return Returns a string that conforms to the StatsD protocol:
   *         KEY:VALUE|METRIC or KEY:VALUE|METRIC|@SAMPLE_RATE
   */
  def stat(key: String, value: String, metric: String, sampleRate: Double) = {
    val sampleRateString = if (sampleRate < 1) "|@" + sampleRate else ""
    key + ":" + value + "|" + metric + sampleRateString
  }
}

/**
 * Message for the StatsDActor
 */
private case class SendStat(stat: String)

/**
 * @param host The statsd host
 * @param port The statsd port
 * @param multiMetrics If true, multiple stats will be sent in a single UDP packet
 * @param packetBufferSize If multiMetrics is true, this is the max buffer size before sending the UDP packet
 */
private class StatsDActor(host: String,
                          port: Int,
                          multiMetrics: Boolean,
                          packetBufferSize: Int) extends Actor {

  private val log = LoggerFactory.getLogger(getClass())

  private val sendBuffer = ByteBuffer.allocate(packetBufferSize)

  private val address = new InetSocketAddress(InetAddress.getByName(host), port)
  private val channel = DatagramChannel.open()

  def receive = {
    case msg: SendStat => doSend(msg.stat)
    case _             => log.error("Unknown message")
  }

  override def postStop() = {
    //save any remaining data to StatsD
    flush

    //Close the channel
    if (channel.isOpen()) {
      channel.close()
    }

    sendBuffer.clear()
  }

  private def doSend(stat: String) = {
    try {
      val data = stat.getBytes("utf-8")

      // If we're going to go past the threshold of the buffer then flush.
      // the +1 is for the potential '\n' in multi_metrics below
      if (sendBuffer.remaining() < (data.length + 1)) {
        flush
      }

      // multiple metrics are separated by '\n'
      if (sendBuffer.position() > 0) {
        sendBuffer.put('\n'.asInstanceOf[Byte])
      }

      // append the data
      sendBuffer.put(data) 

      if (!multiMetrics) {
        flush
      }

    }
    catch {
      case e: IOException => {
        log.error("Could not send stat {} to host {}:{}", sendBuffer.toString, address.getHostName(), address.getPort().toString, e)
      }
    }
  }

  private def flush(): Unit = {
    try {
      val sizeOfBuffer = sendBuffer.position()

      if (sizeOfBuffer <= 0) {
        // empty buffer
        return
      }

      // send and reset the buffer 
      sendBuffer.flip()
      val nbSentBytes = channel.send(sendBuffer, address)
      sendBuffer.limit(sendBuffer.capacity())
      sendBuffer.rewind()

      if (sizeOfBuffer != nbSentBytes) {
        log.error("Could not send entirely stat {} to host {}:{}. Only sent {} bytes out of {} bytes", sendBuffer.toString(),
          address.getHostName(), address.getPort().toString, nbSentBytes.toString, sizeOfBuffer.toString)
      }

    }
    catch {
      case e: IOException => {
        log.error("Could not send stat {} to host {}:{}", sendBuffer.toString, address.getHostName(), address.getPort().toString, e)
      }
    }
  }
}